File: framework/languages/java/e2e/AutomaticTurnDispatch/Client/src/main/java/systems/zlink/e2e/automaticturn/client/Scenarios/AtdA3ContinuationContextScenario.java
package systems.zlink.e2e.automaticturn.client.Scenarios;

import systems.zlink.e2e.automaticturn.client.Support.AutomaticTurnDispatchScenarioSupport;
import systems.zlink.stream.connector.ZLinkStreamConnector;

public final class AtdA3ContinuationContextScenario {
    private AtdA3ContinuationContextScenario() {
    }

    public static void run(ZLinkStreamConnector connector) throws Exception {
        AutomaticTurnDispatchScenarioSupport.runContinuationContext(connector);
        System.out.println("scenario ATD-A3 passed");
    }
}

File: framework/languages/node/e2e/AutomaticTurnDispatch/Client/Scenarios/atd-a3-continuation-context-scenario.ts
// ATD-A3: await 전후 request와 correlation 문맥 보존 검증한다.
import type {
  AwaitMsg,
  AwaitEvidenceRes,
  AwaitEvidenceWaitReq
} from '../../Shared/messages';
import { AutomaticTurnDispatchNames } from '../../Shared/messages';
import { containsRequestMarkersInOrder } from '../Support/scenario-assert';
import type { ZlinkStreamConnector } from '@zlink-systems/stream-connector';

export async function runYdA3(client: ZlinkStreamConnector, spotRid: string): Promise<string> {
  const requestId = `ATD-A3-${uniqueId()}`;
  await client
    .send({ requestId, delayMs: 50, correlationId: 'corr-a3' } satisfies AwaitMsg)
    .packetName('AwaitMsg')
    .metadata(AutomaticTurnDispatchNames.spotRidMetadata, spotRid)
    .submit();

  const evidence = await client
    .request({ requestId, marker: 'await-completed', timeoutMilliseconds: 30000 } satisfies AwaitEvidenceWaitReq)
    .packetName('AwaitEvidenceWaitReq')
    .metadata(AutomaticTurnDispatchNames.targetNodeRidMetadata, 'play-a')
    .timeout(30000)
    .submit<AwaitEvidenceRes>();
  containsRequestMarkersInOrder(evidence.evidence, requestId, [
    'await-started',
    'await-released',
    'await-resumed',
    'await-completed'
  ], 'ATD-A3 marker order mismatch.');
  console.log('scenario ATD-A3 passed');
  return requestId;
}

function uniqueId(): string {
  return `${Date.now().toString(16)}-${Math.random().toString(16).slice(2)}`;
}
